### PR TITLE
Tags for motions

### PR DESCRIPTION
**Pull request by ostcar**
The backend is done. You can alter the tags in the general config tab.

The tags are created in the core app, so we can use the same tabs for assignments or other apps later.

@emanuelschuetze I put the tags as list inside the MotionDetailView as example. Can you make it pretty?

### CODE REVIEW

File: openslides/motion/forms.py [hunk 1]
```diff
@@ -49,6 +49,11 @@ class BaseMotionForm(CleanHtmlFormMixin, CssClassMixin, forms.ModelForm):
     Attachments of the motion.
     """
 
+    tags = forms.ModelMultipleChoiceField(
+        queryset=Tag.objects.all(),
+        required=False,
+        label=ugettext_lazy('Tags'))
+
```
Comment (emanuelschuetze) [inline openslides/motion/forms.py]: Is this code block really required if you use `openslides.core.models import Tag`? For motion you use this in other way than in agenda and assignment.

File: openslides/motion/forms.py [hunk 2]
```diff
@@ -66,6 +71,7 @@ def __init__(self, *args, **kwargs):
             self.initial['text'] = last_version.text
             self.initial['reason'] = last_version.reason
             self.initial['attachments'] = self.motion.attachments.all()
+            self.initial['tags'] = self.motion.tags.all()
```
Comment (emanuelschuetze) [inline openslides/motion/forms.py]: Same here: Is this really required in motion form only?

File: openslides/motion/forms.py (hunk 1 above)
Comment (ostcar) [inline openslides/motion/forms.py]: Yes. The motion form is generated automaticly from different Form-Classes. So it does not work the "normal" way

File: openslides/motion/forms.py (hunk 2 above)
Comment (ostcar) [inline openslides/motion/forms.py]: yes, see above

File: openslides/agenda/models.py [hunk 3]
```diff
@@ -109,6 +110,11 @@ class Item(SlideMixin, AbsoluteUrlMixin, MPTTModel):
     True, if the list of speakers is closed.
     """
 
+    tags = models.ManyToManyField(Tag, blank=True)
+    """
+    Tags to categories agenda items.
```
Comment (normanjaeckel) [inline openslides/agenda/models.py]: Typo: `categorize` (American) or `categorise` (British)

File: openslides/core/models.py [hunk 4]
```diff
@@ -38,3 +38,21 @@ def get_absolute_url(self, link='update'):
         else:
             url = super(CustomSlide, self).get_absolute_url(link)
         return url
+
+
+class Tag(AbsoluteUrlMixin, models.Model):
+    """
+    Model to save tags.
+    """
+    # At the moment there are only tags for motions
```
Comment (normanjaeckel) [inline openslides/core/models.py]: Remove this line.

File: openslides/core/models.py [hunk 5]
```diff
@@ -38,3 +38,21 @@ def get_absolute_url(self, link='update'):
...
+
+    name = models.CharField(max_length=255, unique=True,
+                            verbose_name=ugettext_lazy('Tag'))
+
+    class Meta:
+        ordering = ['name']
+        permissions = (
+            ('can_manage_tags', ugettext_noop('Can create and manipulate tags')), )
```
Comment (normanjaeckel) [inline openslides/core/models.py]: Proposal for the human readable string: `Can manage tags`.

File: openslides/core/templates/core/tag_list.html [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+                        </a>
+                    </span>
+                </td>
+            </tr>
+        {% endfor %}
+    </table>
+{% endblock %}
+
+{% block javascript %}
+<script src="{% static 'js/config_tags.js' %}" type="text/javascript"></script>
```
Comment (normanjaeckel) [inline openslides/core/templates/core/tag_list.html]: Please indent this line with 4 spaces.

File: openslides/core/views.py [hunk 7]
```diff
@@ -213,3 +214,85 @@ class CustomSlideDeleteView(CustomSlideViewMixin, utils_views.DeleteView):
     Delete a custom slide.
     """
     pass
+
+
+class TagException(Exception):
```
Comment (normanjaeckel) [inline openslides/core/views.py]: Can you use a subclass of OpenSlidesError? Can you move this definition into a separate file `exceptions.py`?

File: openslides/core/views.py [hunk 8]
```diff
@@ -213,3 +214,85 @@ class CustomSlideDeleteView(CustomSlideViewMixin, utils_views.DeleteView):
...
+
+    def ajax_get(self, request, *args, **kwargs):
+        name, value = request.POST['name'], request.POST.get('value', None)
+
+        # Create a new tag
+        if name == 'new':
+            try:
+                tag = Tag.objects.create(name=value)
+            except IntegrityError:
+                # The tag-name is already taken (it must be uniquei)
```
Comment (normanjaeckel) [inline openslides/core/views.py]: Typo: ~~# The tag-name is already taken (it must be uniquei)~~ # The name of the tag is already taken. It must be unique.

File: openslides/core/views.py [hunk 9]
```diff
@@ -213,3 +214,85 @@ class CustomSlideDeleteView(CustomSlideViewMixin, utils_views.DeleteView):
...
+            except TagException as error:
+                self.error = str(error)
+            except IntegrityError:
+                self.error = 'Tag name is already taken'
+            except Tag.DoesNotExist:
+                self.error = 'Tag does not exist'
+            else:
+                self.action = 'updated'
+
+        # Delete an tag
```
Comment (normanjaeckel) [inline openslides/core/views.py]: Typo: a~~n~~

File: openslides/core/views.py [hunk 10]
```diff
@@ -213,3 +214,85 @@ class CustomSlideDeleteView(CustomSlideViewMixin, utils_views.DeleteView):
...
+
+    def get_tag_queryset(self, name, place_in_str):
+        """
+        Get a django-tag-queryset from a string.
+
+        'name' is the string in which the pk is (at the end).
+
+        'place_in_str' is the place where to look for the pk. It has to be an int.
+
+        Returns a Tag-queryset or raises TagException.
```
Comment (normanjaeckel) [inline openslides/core/views.py]: Typo: ~~Tag-queryset~~ Tag QuerySet

File: openslides/motion/forms.py [hunk 11]
```diff
@@ -11,7 +11,7 @@
 
 from ckeditor.widgets import CKEditorWidget
 
-from .models import Category, Motion, Workflow
+from .models import Category, Motion, Workflow, Tag
```
Comment (normanjaeckel) [inline openslides/motion/forms.py]: Change order: `from .models import Category, Motion, Tag, Workflow`

File: openslides/motion/models.py [hunk 12]
```diff
@@ -71,8 +72,10 @@ class Motion(SlideMixin, AbsoluteUrlMixin, models.Model):
     Many to many relation to mediafile objects.
     """
 
-    # TODO: proposal
-    # master = models.ForeignKey('self', null=True, blank=True)
+    tags = models.ManyToManyField(Tag)
+    """
+    Tags to categories motions.
```
Comment (normanjaeckel) [inline openslides/motion/models.py]: Typo: `categorize` (American) or `categorise` (British)

File: openslides/participant/templates/participant/edit.html [hunk 13]
```diff
@@ -4,22 +4,7 @@
 {% load staticfiles %}
 
 {% block header %}
```
Comment (normanjaeckel) [inline openslides/participant/templates/participant/edit.html]: The empty block `header` can be removed.

**Comment by normanjaeckel** [comment]
Some remarks. Looks nice but I have not checked the javascript code.

File: openslides/agenda/models.py (hunk 3 above)
Comment (ostcar) [inline openslides/agenda/models.py]: done

File: openslides/core/models.py (hunk 4 above)
Comment (ostcar) [inline openslides/core/models.py]: done

File: openslides/core/models.py (hunk 5 above)
Comment (ostcar) [inline openslides/core/models.py]: done

File: openslides/core/templates/core/tag_list.html (hunk 6 above)
Comment (ostcar) [inline openslides/core/templates/core/tag_list.html]: done

File: openslides/core/views.py (hunk 7 above)
Comment (ostcar) [inline openslides/core/views.py]: done

File: openslides/core/views.py (hunk 8 above)
Comment (ostcar) [inline openslides/core/views.py]: done

File: openslides/core/views.py (hunk 9 above)
Comment (ostcar) [inline openslides/core/views.py]: done

File: openslides/core/views.py (hunk 10 above)
Comment (ostcar) [inline openslides/core/views.py]: done

File: openslides/motion/forms.py (hunk 11 above)
Comment (ostcar) [inline openslides/motion/forms.py]: I do this in another pull request where we integrate isort and sort all imports. See #1383

File: openslides/motion/models.py (hunk 12 above)
Comment (ostcar) [inline openslides/motion/models.py]: done

File: openslides/participant/templates/participant/edit.html (hunk 13 above)
Comment (ostcar) [inline openslides/participant/templates/participant/edit.html]: done

**Comment by normanjaeckel** [comment]
Ok.

**Comment by emanuelschuetze** [comment]
I'm still working on some template improvements...

**Comment by emanuelschuetze** [comment]
Please merge #1380 before that I can use the template changes (of amendment tag) in motion list view.

**Comment by ostcar** [comment]
@emanuelschuetze this branch is now rebased on top of #1380. But please let us merge this first into master. You can put your template improvements in another pull request. This one does not break the master so it can be merged first.
